gr-picoscope/picoscope.hpp

```diff
@@ -0,0 +1,559 @@
+    trigger_setting_t  trigger;
+};
+
+struct State {
+    std::vector<Channel>        channels;
+    std::atomic<std::size_t>    data_available     = 0;
+    std::atomic<bool>           data_finished      = false;
+    bool                        initialized        = false;
+    bool                        configured         = false;
+    bool                        closed             = false;
+    bool                        armed              = false;
+    bool                        started            = false; // TODO transitional until nodes have state handling
+    int16_t                     handle             = -1;    ///< picoscope handle
+    int16_t                     overflow           = 0;     ///< status returned from getValues
+    int16_t                     max_value          = 0;     ///< maximum ADC count used for ADC conversion
+    double                      actual_sample_rate = 0;
+    std::thread                 poller;
+    BufferHelper<Error>         errors;
+    std::atomic<poller_state_t> poller_state = poller_state_t::IDLE;
+    std::atomic<bool>           forced_quit  = false; // TODO transitional until we found out what
+                                                      // goes wrong with multithreaded scheduler
+    std::size_t produced_worker = 0;                  // poller/callback thread
+};
+
+} // namespace detail
+
+// optional shortening
+template<typename T, fair::meta::fixed_string description = "", typename... Arguments>
+using A = fair::graph::Annotated<T, description, Arguments...>;
+
+using fair::graph::Visible;
+
+template<typename PSImpl>
+struct Picoscope : public fair::graph::node<PSImpl, fair::graph::BlockingIO<true>> {
+    A<std::string, "serial number", Visible> serial_number;
+    A<double, "sample rate", Visible>        sample_rate = 10000.;
+    // TODO any way to get custom enums into pmtv??
+    A<std::string, "acquisition mode as string", Visible>      acquisition_mode_string  = std::string("STREAMING");
+    A<std::size_t, "pre-samples", Visible>                     pre_samples              = 1000;
+    A<std::size_t, "post-samples", Visible>                    post_samples             = 9000;
+    A<std::size_t, "no. captures (rapid block mode)", Visible> rapid_block_nr_captures  = 1;
+    A<bool, "trigger once (rapid block mode)", Visible>        trigger_once             = false;
+    A<double, "poll rate (streaming mode)", Visible>           streaming_mode_poll_rate = 0.001;
+    A<bool, "auto-arm", Visible>                               auto_arm                 = true;
+    A<ChannelMap, "enabled analog channels", Visible>          enabled_channels;
+    A<PortMap, "enabled digital ports", Visible>               enabled_ports;
+    A<trigger_setting_t, "trigger settings", Visible>          trigger;
+
+    detail::State                                              state;
+    detail::Settings                                           ps_settings;
+
```

Comment:
Missing settings are -- mostly per port and emitted as tags on the very first sample and whenever these changes:
 * [ ] signal_name(s) (store as std::array)
 * [ ] signal_unit(s)
 * [ ] signal scaling factor (needed for int -> float scaling before samples are emitted)
 * [ ] signal offset (needed for int -> float scaling before samples are emitted)
 
 Additional tags to be emitted are:
  * [ ] signal_min (depends on ADC n_max bits, scaling, and offset)
  * [ ] signal_max (depends on ADC n_max bits, scaling, and offset)
  * [ ] trigger_name (to be emitted when an edge was detected on the trigger port)
  * [ ] trigger_time (to be emitted when an edge was detected on the trigger port)
  * [ ] trigger_offset (to be emitted when an edge was detected on the trigger port)
  
We also need to think how to 
 * [ ] integrate the 'reset()' functionality.

Reply:
Is this for this PR? I agree that the tag handling is missing. Trigger tag handling should be re-added together with the timing message handling (requires message ports)

Reply:
I've added the default tags SAMPLE_RATE, SIGNAL_NAME, SIGNAL_UNIT, SIGNAL_MIN, SIGNAL_MAX.
Trigger tags I would add when readding the trigger logic (i.e. when the message port is available)

Reply:
Please add the trigger tag with a default name. We need this in any case. 

The mapping to the external trigger information could be added now as well. Maybe add a `MsgPortIn timing' now. I will add a default tag definition for the FAIR specific timing info. For the time being assume that this would be called `trigger_meta_information` and be of a type `property_map = pmtv::map_t`.

The Timing receiver would then emit a tag_t that is routed to the 'timing' port and contain the following information:
 * trigger_name
 * trigger_time
 * trigger_offset
 * trigger_meta_information
 * context
 
 The definition are available in the tag.hpp header.

Reply:
Created #126